### ISSUE TITLE
Extend opacus.DPLSTM to support PackedSequences

### ISSUE THREAD

**Issue by Darktex** [pytorch/opacus#103]
## 🚀 Feature
Extend [opacus.DPLSTM](https://github.com/pytorch/opacus/blob/master/opacus/layers/dp_lstm.py) to work with [PackedSequences](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.PackedSequence.html).

This is a good first issue to contribute, and we would very much welcome a PR!

## Motivation

The `PackedSequence` format allows us to minimize padding in a batch by "zipping" sequences together, and keeping track of the lengths. It is a very commonly-used format for torch.nn.LSTM and it is the only feature that our DPLSTM reimplementation does not yet support. 

## Pitch

A common problem in NLP (or in dealing with sequences of stuff, more in general) is uneven lengths of examples in a batch. For example, if you want to batch `short sentence` and `very long long long long sentence` together in a single tensor, you need to handle the fact that the first sequence has length of 2 and the second has length of 6. Matrices (and tensors) can't have uneven rows :)

A common approach to this is simply padding: add a special token, eg `<pad>` to all sequences as needed to reach the length of the longest sequence in the batch (eg `short sentence <pad> <pad> <pad> <pad>`). This works, but leads to wasted memory and computation (the LSTM will still run through the pads). A better approach is having a format such as `PackedSequence`, which instead stacks and zips the sequences, and keeps track of all the relative lengths (see ***Additional Context** for more info on how this is done). This way, we don't waste neither memory nor compute. In our example before, we would pack our two sequences together like this:

`[short, very, sentence, long, long, long, long, sentence]` with `lens=[2, 6]`.

So we "zip" them, and then we just keep going once the shorter sequence is finished. Why do we zip them? Think about how the LSTM runs: it goes step by step, but you can batch every step! This way, we start by running on both `short` and `very` (first items in each seq), then we continue to `sentence` and `long`, and then we can return the result for the first sequence but we keep going for the second sequence (we know how to do this because we have `lens` to tell us).

## Additional context

See more tutorials about PackedSequences:

1. https://gist.github.com/HarshTrivedi/f4e7293e941b17d19058f6fb90ab0fec
2. https://discuss.pytorch.org/t/understanding-pack-padded-sequence-and-pad-packed-sequence/4099/6
3. This StackOverflow [answer](https://stackoverflow.com/a/56211056/371613) has a good visual explanation.

**Comment by adriansarstedt** [comment]
I'm excited to get started with this issue. Will keep regular progress updates on this thread.

**Comment by touqir14** [comment]
@adriansarstedt , I actually started working on this a week back but it got delayed due to my other project works. Since you have started working on this, can we instead collaborate on any potential followup works deemed necessary by the Opacus maintainers?
@Darktex , I just pushed a PR adding support for PackedSequence in DPSLTM. Let me know what you think.

**Comment by touqir14** [comment]
@Darktex , I have submitted a [PR](https://github.com/pytorch/opacus/pull/116) adding PackedSequence support in DPLSTM.
@adriansarstedt , I actually started working on this about a week ago but it got delayed due to my other projects. Since you have started working on this, can we instead collaborate on potential followup works deemed necessary by the opacus maintainers?

**Comment by adriansarstedt** [comment]
Hi @touqir14, shame we couldn't communicate the overlap earlier but it looks like you've done some awesome work! I would be excited to continue working on the issue with you if there are adjustments needed. I've just pushed my first commit to the `packed_sequences_support` branch on my fork. Looks like we've tackled testing the same way which is reassuring. 

While becoming familiar with the project I noted several questions. @Darktex, would you or any other experienced contributors be able to share some insights? 

#### Project Questions:
- Running **black**, **flake8** and **isort** on the project results in 28 formatting changes. Is this consistent with the results on your machine? If not, my local setting must be incorrect.
- Can we include the following clause at the end of tests to enable testing any individual test?

        if __name__ == "__main__":
            unittest.main()

#### dp_layers_test.py Questions:
- The existing `SimpleDPLSTMTest` and `ComplexDPLSTMTest` contain 3 differences in 147 lines, would it be worth unifying the common logic here? This should be achievable by looping with different parameters.

**Comment by Darktex** [comment]
Done!
